### PR TITLE
Sync instrument across components in the reflectometry GUI

### LINKED ISSUE #26536: Sync instrument across all batches
To avoid potential confusion/misuse, we should sync the instrument across all batches in the ISIS Reflectometry GUI and disable changing the instrument in any batch if any batch is processing or autoprocessing. This will need coordinating via the MainWindow similarly to the way the autoprocess button is disabled/enabled.

This is also required for the Slit Calculator.

- [x] Move `m_instrument` and related getter/setter functions from the `BatchPresenter` to the `MainWindow`.
- [x] Add notifications to update the main window when the instrument is changed. Once changed, this should then call down to update all batch presenters with the new instrument.

### PR DESCRIPTION
**Pull request by gemmaguest**
This PR adds functionality to synchronise the instrument across all Batch tabs in the ISIS Reflectometry GUI:
- Add notifications to the MainWindowPresenter when a request to change the instrument is made.
- Move loading/storing the instrument from the BatchPresenter to the MainWindowPresenter.
- Also add notifications to allow updating the instrument without changing the name (this is required when restoring defaults because we want to re-load the instrument).
- Make a minor improvement to the RunsPresenter to only reset the search results when the instrument has been successfully changed.
- Disable changing the Instrument if processing or autoprocessing is running.
- Rename notification functions to be more clear and consistent.
- Remove some unused code regarding the SlitCalculator.

**To test:**

Should be tested by someone at ISIS.

- Open the ISIS Reflectometry interface
- Check that the instrument is the same in the top left and bottom right drop-down boxes on all Batch tabs.
- Change the instrument and check it updates all tabs.
- Try adding/removing batch tabs and check the instrument is the same as the other tabs.
- Try changing some settings on the Experiment and Instrument tabs and then hit Restore Defaults and check they revert to original values. They should also revert to the relevant defaults when you change the instrument.
- On the Runs tab, enter investigation ID `1120015` and click Search. Log in to ICat if requested. Now change the instrument and check that the search ID and results are cleared.
- Click Search again and select all runs and click Transfer. Then click Process to process all runs. Check that the Instrument drop-down is disabled on all Batch tabs (including new ones).
- Click Pause and check that the drop-down is re-enabled on all tabs when processing is finished.
- Click Autoprocess and check the same behaviour. When all runs finish processing the drop-down should still remain disabled (because autoprocessing is polling for new runs) - it should only be re-enabled when you Pause autoprocessing.

Fixes #26536 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/scientific_interfaces/test/ISISReflectometry/Batch/BatchPresenterTest.h [hunk 1]
```diff
@@ -54,16 +54,42 @@ class BatchPresenterTest : public CxxTest::TestSuite {
...
+
+  void testMainPresenterUpdatedWhenChangeInstrumentRequested() {
     auto presenter = makePresenter();
     auto const instrument = std::string("POLREF");
-    EXPECT_CALL(*m_runsPresenter, instrumentChanged(instrument)).Times(1);
-    EXPECT_CALL(*m_experimentPresenter, instrumentChanged(instrument)).Times(1);
-    EXPECT_CALL(*m_instrumentPresenter, instrumentChanged(instrument)).Times(1);
+    EXPECT_CALL(m_mainPresenter, notifyChangeInstrumentRequested(instrument))
+        .Times(1);
+    presenter.notifyChangeInstrumentRequested(instrument);
```
Comment (DavidFair) [inline qt/scientific_interfaces/test/ISISReflectometry/Batch/BatchPresenterTest.h]: I might misunderstand the mechanism behind this: but does calling `notifyChangeInstrumentRequested`not implicitly satisfy the mock expectation that it's called, making it always pass?

**Comment by DavidFair** [comment]
Some minor things highlighted by testing:
- Changing instrument will reset the investigation/experiment settings  - I'm sure this is intended (with clearing the investigation results) but it could be irritating if you have filled in all the details then realise you selected the wrong instrument? Especially when there is a reset to defaults button

- Pause it disabled if you hit process in the bottom right, this means if didn't hit auto process by accident and have lots of runs your fully committed 

- Changing the instrument does not clear the runs table for the current tab - although they process correctly using the original instrument. It's a bit confusing but I can't suggest a way to fix this (if users change instrument on another tab it shouldn't clear)

- A really minor nitpick, the add row button in experiment settings is not disabled whilst processing runes, despite the rows added being disabled

File: qt/scientific_interfaces/test/ISISReflectometry/Batch/BatchPresenterTest.h (hunk 1 above)
Comment (gemmaguest) [reply qt/scientific_interfaces/test/ISISReflectometry/Batch/BatchPresenterTest.h]: Ah, I think that's been left in by mistake - I'll remove it.

**Comment by gemmaguest** [comment]
> * Changing instrument will reset the investigation/experiment settings  - I'm sure this is intended (with clearing the investigation results) but it could be irritating if you have filled in all the details then realise you selected the wrong instrument? Especially when there is a reset to defaults button

The settings are not very "transferable" so if you change instrument I think you typically want to start afresh from that instrument's defaults. It might be nice to have some way to go back, though (either remember the last settings for each instrument so they can return to them if they've toggled instrument by mistake, or have some kind of undo functionality). Users don't tend to change the instrument though so I doubt this will be much of a problem.

> * Pause is disabled if you hit process in the bottom right, this means if you didn't hit auto process rather than process your fully committed to finishing this current batch

Processing can be paused using the pause button on the toolbar rather than the Autoprocess pause. It's a bit confusing having multiple buttons to start/stop processing so we could perhaps consolidate them.

> * Changing the instrument does not clear the runs table for the current tab - although they process correctly using the original instrument. It's still a bit confusing, but I can't suggest a way to fix this (if users change instrument on another tab it shouldn't clear)

I'm in two minds about whether to clear the table. I think it's ok to clear the settings because users are unlikely to start entering settings for the "wrong" instrument, but they might start entering runs before they realise they have the wrong instrument, and then it would be annoying to lose them. I'm tempted to leave it, but it is odd it processes with the original instrument - that sounds like a bug we should fix.

> * A really minor nitpick, the add row button in experiment settings is not disabled whilst processing runs, despite the rows added being disabled

Good spot - that should be an easy fix :+1:

**Comment by gemmaguest** [comment]
I've fixed the conflicts and I'll take a look at the processing-with-previous-instrument problem and remove the spurious EXPECT_CALL. The other issues are perhaps best looked at as separate issues.

**Comment by gemmaguest** [comment]
The EXPECT_CALL is for a call to the parent presenter, so I think this is actually a valid test.

The processing-with-previous-instrument problem I also think is ok, unless I've missed something. It doesn't actually use the previous instrument unless the workspace has already been loaded, in which case it uses the already-loaded workspace, which is the expected behaviour of the underlying algorithm. It might be good to add some validation that we have the correct instrument, but that's beyond the scope of this PR.

I spotted an unrelated problem with the BatchPresenterTests - these were crashing on my Windows machine due to a move constructor being called unexpectedly. I've added a fix for that.

**Comment by gemmaguest** [comment]
@DavidFair I think this is ok to go ahead now but let me know if I missed anything you think should be in this PR

**Review by DavidFair — APPROVED** [review]
I'm happy with these changes - a large proportion of this pull request is renaming of base and derived methods to much clearer names.
The remaining other changes to sync instruments work correctly, and the few niggles picked up during testing will be addressed in future issues, as they aren't directly related to this PR.

So I'm happy to :shipit:

**Pull request merged**